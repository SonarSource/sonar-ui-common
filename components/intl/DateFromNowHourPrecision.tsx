/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { differenceInHours } from 'date-fns';
import * as React from 'react';
import { DateSource } from 'react-intl';
import { translate } from '../../helpers/l10n';
import Tooltip from '../controls/Tooltip';
import DateFromNow from './DateFromNow';
import DateTimeFormatter from './DateTimeFormatter';

export interface DateFromNowHourPrecisionProps {
  children?: (formattedDate: string) => React.ReactNode;
  date?: DateSource;
}

export default function DateFromNowHourPrecision(props: DateFromNowHourPrecisionProps) {
  const { children = (f: string) => f, date } = props;

  /*
   * Cast returns as ReactElement to bypass typescript issue with functional components
   * returning anything else (like a string, in our case...)
   * (https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20544)
   */
  if (!date) {
    return children(translate('never')) as React.ReactElement;
  }

  if (differenceInHours(Date.now(), date) < 1) {
    return children(translate('less_than_1_hour_ago')) as React.ReactElement;
  }

  return (
    <Tooltip overlay={<DateTimeFormatter date={date} />}>
      <span>
        <DateFromNow date={date}>{children}</DateFromNow>
      </span>
    </Tooltip>
  );
}
