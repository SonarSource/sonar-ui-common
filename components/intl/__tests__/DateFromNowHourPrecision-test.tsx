/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import DateFromNowHourPrecision, {
  DateFromNowHourPrecisionProps,
} from '../DateFromNowHourPrecision';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('standard');
  expect(shallowRender({ date: undefined }, true)).toMatchSnapshot(
    'overridden - no date no children'
  );
  expect(shallowRender({ date: new Date() })).toMatchSnapshot('overridden - less than an hour');
});

function shallowRender(overrides: Partial<DateFromNowHourPrecisionProps> = {}, noChildren = false) {
  const date = new Date('2020-02-20T20:20:20Z');
  return noChildren
    ? shallow(<DateFromNowHourPrecision date={date} {...overrides} />)
    : shallow(
        <DateFromNowHourPrecision date={date} {...overrides}>
          {(formatted) => <span>{formatted}</span>}
        </DateFromNowHourPrecision>
      );
}
