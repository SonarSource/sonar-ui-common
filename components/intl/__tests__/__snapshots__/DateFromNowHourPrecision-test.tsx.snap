// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: overridden - less than an hour 1`] = `
<span>
  less_than_1_hour_ago
</span>
`;

exports[`should render correctly: overridden - no date no children 1`] = `"never"`;

exports[`should render correctly: standard 1`] = `
<Tooltip
  overlay={
    <DateTimeFormatter
      date={2020-02-20T20:20:20.000Z}
    />
  }
>
  <span>
    <DateFromNow
      date={2020-02-20T20:20:20.000Z}
    >
      <Component />
    </DateFromNow>
  </span>
</Tooltip>
`;
